sklearn/preprocessing/_encoders.py

```diff
@@ -97,7 +97,27 @@ def _fit(
                 else:
                     cats = result
             else:
-                cats = np.array(self.categories[i], dtype=Xi.dtype)
+                if np.issubdtype(Xi.dtype, np.str_):
+                    # Always convert string categories to objects to avoid
+                    # unexpected string truncation for longer category labels
+                    # passed in the constructor.
+                    Xi_dtype = object
+                else:
+                    Xi_dtype = Xi.dtype
+
+                cats = np.array(self.categories[i], dtype=Xi_dtype)
+                if (
+                    cats.dtype == object
+                    and isinstance(cats[0], bytes)
+                    and Xi.dtype.kind != "S"
+                ):
+                    msg = (
+                        f"Found incompatible types ('{type(cats[0]).__name__}' vs"
+                        f" '{type(Xi[0]).__name__}') between values and predefined"
+                        f" categories in column {i}."
+                    )
```

Comment:
I think this is a slightly clearer error message. `'bytes'` can be hard coded because of the `if` statement above it:

```suggestion
                    msg = (
                        f"In column {i}, the predefined categories has type"
                        " 'bytes' which is incompatible with"
                        f" values of type '{type(Xi[0]).__name__}'."
                    )
                    raise ValueError(msg)
```

Although, using bytes as predefined categories is an edge case, but this PR is a backward incompatible change. If we are okay with it, then it needs to be in the "Changed Model" section of the whats new.

Reply:
I prefer your message. Much less clunky than what I came up with. Also moved the change log entry.